### DISCUSSION TITLE
Проверка требования аутентификации

### DISCUSSION THREAD

**Discussion by ivanych** [schemathesis/schemathesis#1273] [Q&A]
Хочу проверить, что все методы, которые в спеке указаны как требующие аутентификацию, действительно её требуют.

Кажется, для этого было бы достаточно такой проверки:

"Если вызов этого метода без выполнения какой-либо аутентификации вернул код ответа, отличный от 401, то проверка провалена".

Можно ли такое сделать? Как?

**Comment by ivanych** [comment]
@Stranger6667 есть ли решение? Может опция какая?

**Reply by Stranger6667** [reply]
Пардон, не заметил сообщение - постараюсь ответить сегодня / завтра

**Reply by Stranger6667** [reply]
> "Если вызов этого метода без выполнения какой-либо аутентификации вернул код ответа, отличный от 401, то проверка провалена".
> Можно ли такое сделать? Как?

Для Python тестов оно примерно в этом ключе и будет:

```python
from hypothesis import settings

...

def requires_auth(operation):
    # If it is present and non-empty then API operation requires some auth
    # You might go deeper and filter only some auth options if needed (e.g. only api key, etc)
    security = operation.definition.resolved.get("security")
    if security is not None:
        # Operation-level "security" overrides the schema-level one
        return bool(security)
    return bool(operation.schema.raw_schema.get("security"))

@schema.parametrize()
@settings(max_examples=1)  # not likely it requires more than 1 request to check
def test_auth(case):
    # Check whether security is defined for the current API operation
    if requires_auth(case.operation):
        # Drop auth
        if case.headers:
            case.headers.pop("Authorization", None)   # Or any other auth-related header you use
        response = case.call()
        assert response.status_code == 401
```

Для CLI может быть несколько вариантов. Например это можно реализовать в виде [кастомной проверки](https://schemathesis.readthedocs.io/en/stable/cli.html#registering-custom-checks) + хука [add_case](https://schemathesis.readthedocs.io/en/stable/extending.html#add-case). Т.е. хук бы генерировал копию тест кейса только убирал бы из него аутентификацию, потом внутри проверки можно было бы проверять если получен 401.

Сам код немного посложнее будет чем для обычных Python тестов, если того примера не хватит, то напишу полный код для CLI :) Дай знать

**Reply by ivanych** [reply]
Честно сказать, ожидал, что такая проверка есть "из коробки" и включается комбинацией каких-нибудь опций:)

Вроде это напрашивающаяся проверка: аутентификация - совершенно стандартная и частая процедура, такой метод не должен выдавать 200, как и не должен выдавать 500:)

**Reply by Stranger6667** [reply]
Согласен, думаю, что можно такое реализовать и в самом Schemathesis :)
